File: soccer-form-ai/js/biomechanics.js
/* biomechanics.js — pure geometry helpers.
 *
 * All functions operate on "pts": an array of 33 points already converted from
 * MediaPipe's normalized landmark space into ASPECT-CORRECTED pixel space
 * ({x: lm.x * frameW, y: lm.y * frameH, v: visibility}).
 *
 * Why the conversion matters: MediaPipe normalizes x by frame width and y by
 * frame height independently, so on a non-square frame the normalized space is
 * anisotropically stretched. Angles measured directly in that space are wrong.
 * Multiplying back by (W, H) restores a uniform scale so angles and distance
 * ratios are geometrically valid.
 *
 * Everything downstream is expressed as an angle or as a distance divided by
 * torso length, so results are independent of resolution, camera distance and
 * player size.
 */

// MediaPipe Pose landmark indices (33-point model).
export const LM = {
  NOSE: 0,
  L_SH: 11, R_SH: 12, L_EL: 13, R_EL: 14, L_WR: 15, R_WR: 16,
  L_HIP: 23, R_HIP: 24, L_KN: 25, R_KN: 26, L_AN: 27, R_AN: 28,
  L_HEEL: 29, R_HEEL: 30, L_FOOT: 31, R_FOOT: 32
};

// Per-side index bundles so analysis code can be written once and applied to
// whichever leg turns out to be the kicking / plant leg.
export const SIDES = {
  left:  { sh: LM.L_SH, el: LM.L_EL, wr: LM.L_WR, hip: LM.L_HIP, kn: LM.L_KN, an: LM.L_AN, heel: LM.L_HEEL, foot: LM.L_FOOT },
  right: { sh: LM.R_SH, el: LM.R_EL, wr: LM.R_WR, hip: LM.R_HIP, kn: LM.R_KN, an: LM.R_AN, heel: LM.R_HEEL, foot: LM.R_FOOT }
};

// Skeleton edges used for the overlay drawing.
export const CONNECTIONS = [
  [LM.L_SH, LM.R_SH], [LM.L_HIP, LM.R_HIP],
  [LM.L_SH, LM.L_HIP], [LM.R_SH, LM.R_HIP],
  [LM.L_SH, LM.L_EL], [LM.L_EL, LM.L_WR],
  [LM.R_SH, LM.R_EL], [LM.R_EL, LM.R_WR],
  [LM.L_HIP, LM.L_KN], [LM.L_KN, LM.L_AN],
  [LM.L_AN, LM.L_HEEL], [LM.L_HEEL, LM.L_FOOT], [LM.L_AN, LM.L_FOOT],
  [LM.R_HIP, LM.R_KN], [LM.R_KN, LM.R_AN],
  [LM.R_AN, LM.R_HEEL], [LM.R_HEEL, LM.R_FOOT], [LM.R_AN, LM.R_FOOT]
];

export const clamp = (v, lo, hi) => Math.min(hi, Math.max(lo, v));
export const mid = (a, b) => ({ x: (a.x + b.x) / 2, y: (a.y + b.y) / 2, v: Math.min(a.v ?? 1, b.v ?? 1) });
export const dist = (a, b) => Math.hypot(a.x - b.x, a.y - b.y);

/** Convert a MediaPipe landmark list to aspect-corrected pixel points. */
export function toPixels(landmarks, w, h) {
  return landmarks.map(l => ({ x: l.x * w, y: l.y * h, v: l.visibility ?? 1 }));
}

/** Interior angle at vertex `b`, in degrees, formed by a-b-c. 180 = straight. */
export function angleDeg(a, b, c) {
  const v1x = a.x - b.x, v1y = a.y - b.y;
  const v2x = c.x - b.x, v2y = c.y - b.y;
  const n1 = Math.hypot(v1x, v1y), n2 = Math.hypot(v2x, v2y);
  if (n1 < 1e-6 || n2 < 1e-6) return NaN;
  const cos = clamp((v1x * v2x + v1y * v2y) / (n1 * n2), -1, 1);
  return Math.acos(cos) * 180 / Math.PI;
}

export const hipCenter = p => mid(p[LM.L_HIP], p[LM.R_HIP]);
export const shoulderCenter = p => mid(p[LM.L_SH], p[LM.R_SH]);

/**
 * Torso length (hip centre -> shoulder centre) in pixels. This is the unit for
 * every normalized distance in the app: "0.4" means "0.4 torso lengths".
 * Falls back to hip width * 2.2 if the torso is badly foreshortened.
 */
export function torsoScale(p) {
  const t = dist(hipCenter(p), shoulderCenter(p));
  const hipW = dist(p[LM.L_HIP], p[LM.R_HIP]);
  return Math.max(t, hipW * 2.2, 1e-3);
}

/**
 * Signed torso lean from vertical, in degrees.
 * Positive = chest leaning FORWARD along the kick direction; negative = leaning back.
 * `dir` is +1 if the player kicks toward increasing image-x, else -1.
 */
export function torsoLeanDeg(p, dir) {
  const h = hipCenter(p), s = shoulderCenter(p);
  const forward = (s.x - h.x) * dir;   // horizontal displacement toward the target
  const up = h.y - s.y;                // image y grows downward, so this is positive when upright
  return Math.atan2(forward, up) * 180 / Math.PI;
}

/** Signed horizontal offset of point `pt` from `ref`, along the kick direction, in torso units. */
export function forwardOffset(pt, ref, dir, scale) {
  return ((pt.x - ref.x) * dir) / scale;
}

/** Height of `pt` above `ref` (positive = higher in the world), in torso units. */
export function heightAbove(pt, ref, scale) {
  return (ref.y - pt.y) / scale;
}

/** Minimum visibility across the given landmark indices — used as a confidence gate. */
export function minVis(p, idxs) {
  return idxs.reduce((m, i) => Math.min(m, p[i]?.v ?? 0), 1);
}

/**
 * Decide which leg kicks and which way the player faces.
 *
 * Kicking leg: across the contact and follow-through frames, the kicking ankle
 * is the one lifted highest relative to the hips. Summing both frames makes the
 * call robust when contact happens low to the ground.
 *
 * Direction: the kicking ankle travels toward the target from the backswing to
 * the follow-through, so the sign of that horizontal travel gives the facing.
 * If travel is tiny (near head-on camera), fall back to which way the plant
 * foot's toes point relative to its heel.
 */
export function deriveContext(frames) {
  const { plant, contact, followThrough } = frames;
  const lift = (p, side) => {
    const s = SIDES[side];
    return heightAbove(p[s.an], hipCenter(p), torsoScale(p));
  };
  const leftScore = lift(contact.pts, 'left') + lift(followThrough.pts, 'left');
  const rightScore = lift(contact.pts, 'right') + lift(followThrough.pts, 'right');
  const kickSide = leftScore > rightScore ? 'left' : 'right';
  const plantSide = kickSide === 'left' ? 'right' : 'left';

  const kAn = SIDES[kickSide].an;
  const travel = followThrough.pts[kAn].x - plant.pts[kAn].x;
  const scale = torsoScale(contact.pts);
  let dir;
  if (Math.abs(travel) / scale > 0.15) {
    dir = Math.sign(travel);
  } else {
    const ps = SIDES[plantSide];
    const toe = contact.pts[ps.foot].x - contact.pts[ps.heel].x;
    dir = Math.sign(toe) || 1;
  }

  return {
    kickSide, plantSide, dir,
    kick: SIDES[kickSide],
    plant: SIDES[plantSide],
    // How confidently we picked a leg — near-equal lift means a head-on camera.
    legConfidence: Math.min(1, Math.abs(leftScore - rightScore) / 0.5)
  };
}

File: server/biomechanics.py
from __future__ import annotations

import math
from dataclasses import dataclass, field

import numpy as np

X, Y, V = 0, 1, 2


class LM:
    NOSE = 0
    L_SH, R_SH = 11, 12
    L_EL, R_EL = 13, 14
    L_WR, R_WR = 15, 16
    L_HIP, R_HIP = 23, 24
    L_KN, R_KN = 25, 26
    L_AN, R_AN = 27, 28
    L_HEEL, R_HEEL = 29, 30
    L_FOOT, R_FOOT = 31, 32


SIDES = {
    "left":  {"sh": LM.L_SH, "el": LM.L_EL, "wr": LM.L_WR, "hip": LM.L_HIP,
              "kn": LM.L_KN, "an": LM.L_AN, "heel": LM.L_HEEL, "foot": LM.L_FOOT},
    "right": {"sh": LM.R_SH, "el": LM.R_EL, "wr": LM.R_WR, "hip": LM.R_HIP,
              "kn": LM.R_KN, "an": LM.R_AN, "heel": LM.R_HEEL, "foot": LM.R_FOOT},
}

CONNECTIONS = [
    (LM.L_SH, LM.R_SH), (LM.L_HIP, LM.R_HIP),
    (LM.L_SH, LM.L_HIP), (LM.R_SH, LM.R_HIP),
    (LM.L_SH, LM.L_EL), (LM.L_EL, LM.L_WR),
    (LM.R_SH, LM.R_EL), (LM.R_EL, LM.R_WR),
    (LM.L_HIP, LM.L_KN), (LM.L_KN, LM.L_AN),
    (LM.L_AN, LM.L_HEEL), (LM.L_HEEL, LM.L_FOOT), (LM.L_AN, LM.L_FOOT),
    (LM.R_HIP, LM.R_KN), (LM.R_KN, LM.R_AN),
    (LM.R_AN, LM.R_HEEL), (LM.R_HEEL, LM.R_FOOT), (LM.R_AN, LM.R_FOOT),
]


def to_pixels(landmarks, w: int, h: int) -> np.ndarray:
    return np.array(
        [[lm.x * w, lm.y * h, getattr(lm, "visibility", 1.0)] for lm in landmarks],
        dtype=np.float64,
    )


def dist(a, b) -> float:
    return float(math.hypot(a[X] - b[X], a[Y] - b[Y]))


def mid(a, b) -> np.ndarray:
    return np.array([(a[X] + b[X]) / 2, (a[Y] + b[Y]) / 2, min(a[V], b[V])])


def angle_deg(a, b, c) -> float:
    v1 = (a[X] - b[X], a[Y] - b[Y])
    v2 = (c[X] - b[X], c[Y] - b[Y])
    n1, n2 = math.hypot(*v1), math.hypot(*v2)
    if n1 < 1e-6 or n2 < 1e-6:
        return float("nan")
    cos = max(-1.0, min(1.0, (v1[0] * v2[0] + v1[1] * v2[1]) / (n1 * n2)))
    return math.degrees(math.acos(cos))


def hip_center(p: np.ndarray) -> np.ndarray:
    return mid(p[LM.L_HIP], p[LM.R_HIP])


def shoulder_center(p: np.ndarray) -> np.ndarray:
    return mid(p[LM.L_SH], p[LM.R_SH])


def torso_scale(p: np.ndarray) -> float:
    t = dist(hip_center(p), shoulder_center(p))
    hip_w = dist(p[LM.L_HIP], p[LM.R_HIP])
    return max(t, hip_w * 2.2, 1e-3)


def torso_lean_deg(p: np.ndarray, direction: int) -> float:
    h, s = hip_center(p), shoulder_center(p)
    forward = (s[X] - h[X]) * direction
    up = h[Y] - s[Y]
    return math.degrees(math.atan2(forward, up))


def forward_offset(pt, ref, direction: int, scale: float) -> float:
    return ((pt[X] - ref[X]) * direction) / scale


def height_above(pt, ref, scale: float) -> float:
    return (ref[Y] - pt[Y]) / scale


def min_vis(p: np.ndarray, idxs) -> float:
    return float(min(p[i][V] for i in idxs)) if len(idxs) else 0.0


SIDE_ON_SPAN = 0.42
FRONT_ON_SPAN = 0.78


@dataclass
class ViewQuality:
    score: float
    shoulder_ratio: float
    label: str

    def as_dict(self) -> dict:
        ratio = self.shoulder_ratio
        return {"score": round(self.score, 3),
                "shoulderRatio": round(ratio, 3) if math.isfinite(ratio) else None,
                "label": self.label}


def view_quality(p: np.ndarray) -> ViewQuality:
    scale = torso_scale(p)
    span = dist(p[LM.L_SH], p[LM.R_SH]) / scale if scale > 0 else float("nan")
    if not math.isfinite(span):
        return ViewQuality(0.5, float("nan"), "unknown")

    if span <= SIDE_ON_SPAN:
        score = 1.0
    elif span >= FRONT_ON_SPAN:
        score = 0.0
    else:
        score = (FRONT_ON_SPAN - span) / (FRONT_ON_SPAN - SIDE_ON_SPAN)

    label = "side-on" if score >= 0.66 else ("angled" if score >= 0.33 else "face-on")
    return ViewQuality(score, span, label)


@dataclass
class Context:
    kick_side: str
    plant_side: str
    direction: int
    dir_source: str
    leg_confidence: float
    leg_source: str
    view: ViewQuality
    kick: dict = field(default_factory=dict)
    plant: dict = field(default_factory=dict)

    def as_dict(self) -> dict:
        return {
            "kickSide": self.kick_side,
            "plantSide": self.plant_side,
            "dir": self.direction,
            "dirSource": self.dir_source,
            "legConfidence": round(self.leg_confidence, 3),
            "legSource": self.leg_source,
            "view": self.view.as_dict(),
        }


def derive_context(frames: dict, footedness: str = "auto") -> Context:
    plant, contact, follow = frames["plant"], frames["contact"], frames["followThrough"]

    def lift(p, side):
        s = SIDES[side]
        return height_above(p[s["an"]], hip_center(p), torso_scale(p))

    left_score = lift(contact["pts"], "left") + lift(follow["pts"], "left")
    right_score = lift(contact["pts"], "right") + lift(follow["pts"], "right")

    if footedness in ("left", "right"):
        kick_side, leg_source = footedness, "stated"
        leg_confidence = min(1.0, abs(left_score - right_score) / 0.5)
    else:
        kick_side = "left" if left_score > right_score else "right"
        leg_source = "inferred"
        leg_confidence = min(1.0, abs(left_score - right_score) / 0.5)

    plant_side = "right" if kick_side == "left" else "left"

    scale = torso_scale(contact["pts"])
    k_an = SIDES[kick_side]["an"]
    ankle_travel = follow["pts"][k_an][X] - plant["pts"][k_an][X]
    ball_travel = None
    if plant.get("ball") and follow.get("ball"):
        ball_travel = follow["ball"]["x"] - plant["ball"]["x"]

    if ball_travel is not None and abs(ball_travel) / scale > 0.15:
        direction, dir_source = int(math.copysign(1, ball_travel)), "ball"
    elif abs(ankle_travel) / scale > 0.15:
        direction, dir_source = int(math.copysign(1, ankle_travel)), "ankle"
    else:
        ps = SIDES[plant_side]
        toe = contact["pts"][ps["foot"]][X] - contact["pts"][ps["heel"]][X]
        direction = int(math.copysign(1, toe)) if toe else 1
        dir_source = "toe"

    return Context(
        kick_side=kick_side,
        plant_side=plant_side,
        direction=direction,
        dir_source=dir_source,
        leg_confidence=leg_confidence,
        leg_source=leg_source,
        view=view_quality(contact["pts"]),
        kick=SIDES[kick_side],
        plant=SIDES[plant_side],
    )


def inferred_kick_side(frames: dict) -> str:
    def lift(p, side):
        s = SIDES[side]
        return height_above(p[s["an"]], hip_center(p), torso_scale(p))
    left = lift(frames["contact"]["pts"], "left") + lift(frames["followThrough"]["pts"], "left")
    right = lift(frames["contact"]["pts"], "right") + lift(frames["followThrough"]["pts"], "right")
    return "left" if left > right else "right"
